include/swift/AST/ArgumentList.h

```diff
@@ -0,0 +1,649 @@
+
+  /// The argument label written in the call.
+  Identifier getLabel() const { return Label; }
+
+  /// Set a new argument label.
+  ///
+  /// Note this is marked \c & to prevent its use on an rvalue Argument vended
+  /// from an ArgumentList.
+  void setLabel(Identifier newLabel) & { Label = newLabel; }
+
+  /// The argument expression.
+  Expr *getExpr() const { return ArgExpr; }
+
+  /// Set a new argument expression.
+  ///
+  /// Note this is marked \c & to prevent its use on an rvalue Argument vended
+  /// from an ArgumentList.
+  void setExpr(Expr *newExpr) & { ArgExpr = newExpr; }
+
+  /// Whether the argument is \c inout, denoted with the '&' prefix.
+  bool isInOut() const;
+
+  bool operator==(const Argument &other) {
+    return LabelLoc == other.LabelLoc && Label == other.Label &&
+           ArgExpr == other.ArgExpr;
+  }
+  bool operator!=(const Argument &other) { return !(*this == other); }
+};
+
+/// Represents the argument list of a function call or subscript access.
+/// \code
+/// foo.bar(arg, label: arg2, other: &arg3)
+///        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+/// \endcode
+///
+/// Note that arguments are split up and stored in separate buffers to optimize
+/// ArgumentList's memory footprint for the unlabeled cases.
+class alignas(Argument) ArgumentList final
+    : private llvm::TrailingObjects<ArgumentList, Expr *, Identifier,
+                                    SourceLoc> {
+  friend TrailingObjects;
+
+  SourceLoc LParenLoc;
+  SourceLoc RParenLoc;
+
+  /// The number of arguments in the list.
+  unsigned NumArgs : 16;
+
+  /// The index of the first trailing closure, or \c NumArgs if there are no
+  /// trailing closures. In type-checked AST, this may be an index to a variadic
+  /// expansion that contains the first trailing closure.
+  unsigned RawFirstTrailingClosureIndex : 16;
+
+  /// Whether the argument list has multiple trailing closures.
+  bool HasMultipleTrailingClosures : 1;
+
+  /// Whether the argument list was implicitly generated.
+  bool IsImplicit : 1;
+
+  /// Whether any of the arguments are labeled.
+  bool HasLabels : 1;
+
+  /// Whether any of the arguments have labels with location info.
+  bool HasLabelLocs : 1;
+
+  ArgumentList(SourceLoc lParenLoc, SourceLoc rParenLoc, unsigned numArgs,
+               Optional<unsigned> firstTrailingClosureIndex,
+               bool hasMultipleTrailingClosures, bool isImplicit,
+               bool hasLabels, bool hasLabelLocs)
+      : LParenLoc(lParenLoc), RParenLoc(rParenLoc), NumArgs(numArgs),
+        HasMultipleTrailingClosures(hasMultipleTrailingClosures),
+        IsImplicit(isImplicit), HasLabels(hasLabels),
+        HasLabelLocs(hasLabelLocs) {
+    assert(NumArgs == numArgs && "Num args truncated");
+    if (auto idx = firstTrailingClosureIndex) {
+      assert(*idx < numArgs);
+      RawFirstTrailingClosureIndex = *idx;
+    } else {
+      RawFirstTrailingClosureIndex = numArgs;
+      assert(!HasMultipleTrailingClosures);
+    }
```

Comment:
Hmm the if statement felt better to me at the time because of the additional logic, but I don't really feel strongly about it at all, I can change it over